pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+// supports types from the same schema
+func (v *PulumiInputValidator) getType(typeRef string) *pschema.ComplexTypeSpec {
+	if strings.HasPrefix(typeRef, "#/types/") {
+		ref := strings.TrimPrefix(typeRef, "#/types/")
+		if typeSpec, ok := v.schema.Types[ref]; ok {
+			return &typeSpec
+		}
+	}
+	return nil
+}
+
+// ValidateInputs will validate a set of inputs against the pulumi schema. It will
+// return a list of type failures if any are found
+func (v *PulumiInputValidator) ValidateInputs(inputs resource.PropertyMap) *[]TypeFailure {
+	failures := []TypeFailure{}
+	for key, value := range inputs {
+		if failure := v.validateInputType(string(key), value, pathBuilder{paths: []string{}}); failure != nil {
+			failures = append(failures, *failure...)
+		}
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+func (v *PulumiInputValidator) validateInputType(
+	inputName string,
+	inputValue resource.PropertyValue,
+	pathBuilder pathBuilder,
+) *[]TypeFailure {
+	resourceType := v.urn.Type()
+	failures := []TypeFailure{}
+	if resource, ok := v.schema.Resources[resourceType.String()]; ok {
+		if prop, ok := resource.InputProperties[inputName]; ok {
+			if failure := v.validateTypeSpec(
+				inputName,
+				inputValue,
+				prop.TypeSpec,
+				pathBuilder,
+				resource.RequiredInputs,
+			); failure != nil {
+				failures = append(failures, *failure...)
+			}
+		} else {
+			return &[]TypeFailure{
+				{
+					ResourcePath: pathBuilder.toPath(),
+					Reason:       fmt.Sprintf("property %s is not defined in the schema", inputName),
+				},
+			}
+		}
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+type possibleTypes []string
+
+func (p possibleTypes) add(typ ...string) possibleTypes {
+	for _, t := range typ {
+		if !slices.Contains(p, t) && t != "" {
+			p = append(p, t)
+
+		}
+	}
+	return p
+}
+
+func (p possibleTypes) toString() string {
+	return strings.Join(p, " OR ")
+}
+
+// findOneOf attempts to find the correct type spec for a given input input value
+// from a list of possible types
+func (v *PulumiInputValidator) findOneOf(
+	inputValue resource.PropertyValue,
+	specs []pschema.TypeSpec,
+) (*pschema.TypeSpec, []string, bool) {
+
+	allSpecs := []string{}
+	for _, spec := range specs {
+		allSpecs = append(allSpecs, spec.Type)
+		switch inputValue.TypeString() {
+		case "[]":
+			if spec.Type == "array" {
+				if spec.Items != nil {
+					return spec.Items, []string{}, true
+				} else if len(spec.OneOf) > 0 {
+					return v.findOneOf(inputValue, spec.OneOf)
+				}
+				return &spec, []string{}, true
+			}
+		case "object":
+			if spec.Type == "object" {
+				return &spec, []string{}, true
+			}
+		default:
+			if spec.Type == inputValue.TypeString() {
+				return &spec, []string{}, true
+			}
+
+		}
+	}
+	return nil, allSpecs, false
+}
+
+// The input type (i.e. PropertyValue.TypeString() do not match 100% to the pschema.TypeSpec.Type)
+func typeEqual(schemaType, inputType string) bool {
```

Comment:
This is a bit difficult to keep giving the right answer if anything changes in the underlying types, curious if we could get by without it at all.

Reply:
I refactored this to use more of the `PropertyValue` type, but I don't think we can do without this since it's not a 1:1 mapping between the schema types and the `PropertyValue` types.